include/circt/Dialect/Handshake/HandshakeOps.td

```diff
@@ -1249,3 +1249,131 @@ def EndOp : Handshake_Op<"end", [Terminator]> {
     mlir::ValueRange getMemoryControls();
   }];
 }
+
+//===----------------------------------------------------------------------===//
+// Speculation operations
+//===----------------------------------------------------------------------===//
+
+def SpeculatorOp : Handshake_Op<"speculator", [ 
+  TypesMatchWith<"data operand type matches data result type",
+                 "dataIn", "dataOut", "$_self">]> {
+  let summary = "Central control unit of the speculative circuit.";
+  let description = [{
+    The speculator produces speculative tokens when real data has
+    not yet arrived. The speculator is also responsible for deciding 
+    the correctness of its speculation.
+
+    It needs to send control tokens to save units, commit units, 
```

Comment:
I'm assuming these are, respectively, `saveCtrl`, `commitCtrl`, and `scBranchControl`. What are `SCControl1` and `SCControl2` then? Why are they 3 bits? Describing them briefly in this `description` field would be useful. If feasible, also consider using more explicit names, 1 and 2 are not very clear as suffixes.